File: examples/node/docs/privateActiveOrdersServiceStreamActiveOrders.js
/**
 * Copyright 2019 Tulip Solutions B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const util = require('util');

const orders_pb = require('@tulipsolutions/tecl/common/orders_pb');
const order_pb = require('@tulipsolutions/tecl/priv/order_pb');
const order_grpc = require('@tulipsolutions/tecl/priv/order_grpc_pb');

function privateActiveOrdersServiceStreamActiveOrders(host, credentials, options) {
  // CODEINCLUDE-BEGIN-MARKER: ref-code-example-request
  const client = new order_grpc.PrivateActiveOrdersServiceClient(host, credentials);

  // Create a request for streaming all your active orders
  // no fields are set as it does not have any
  const request = new order_pb.StreamActiveOrdersRequest();

  // Make the request asynchronously
  const call = client.streamActiveOrders(request, options);
  call.on('data', response => {
    console.log(response.toObject());
    // CODEINCLUDE-END-MARKER: ref-code-example-request
    parseAndPrint(response);
    // CODEINCLUDE-BEGIN-MARKER: ref-code-example-request
  });
  call.on('error', err => {
    console.error('PrivateActiveOrdersService.StreamActiveOrders error: ' + err.message);
  });
  call.on('end', () => {
    console.log('PrivateActiveOrdersService.StreamActiveOrders completed');
  });
  // CODEINCLUDE-END-MARKER: ref-code-example-request
}

function parseAndPrint(response) {
  // CODEINCLUDE-BEGIN-MARKER: ref-code-example-response
  let orderTypeDetail = '';
  switch (response.getOrderCase()) {
    case order_pb.ActiveOrderStatus.OrderCase.LIMIT_ORDER: {
      const order = response.getLimitOrder();
      orderTypeDetail = util.format(
        '%s %f@%f remaining %f',
        Object.keys(orders_pb.Side).find(key => orders_pb.Side[key] === order.getSide()),
        order.getBaseAmount(),
        order.getPrice(),
        order.getBaseRemaining()
      );
      break;
    }
    // Note that market orders do not show in active orderbook.
    default:
      orderTypeDetail = 'was removed from orderbook';
  }
  let deadline = '(no deadline)';
  if (response.getDeadlineNs() != 0) {
    deadline = util.format('deadline @ %d', response.getDeadlineNs());
  }
  console.log(
    util.format(
      '%s: %s for market %s %s %s',
      'ActiveOrderStatus',
      response.getOrderId(),
      Object.keys(orders_pb.Market).find(key => orders_pb.Market[key] === response.getMarket()),
      orderTypeDetail,
      deadline
    )
  );
  // CODEINCLUDE-END-MARKER: ref-code-example-response
}

module.exports = privateActiveOrdersServiceStreamActiveOrders;

File: examples/java/docs/PrivateActiveOrdersServiceStreamActiveOrders.java
// Copyright 2019 Tulip Solutions B.V.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package docs;

import io.grpc.ManagedChannel;
import io.grpc.stub.StreamObserver;
import nl.tulipsolutions.api.priv.ActiveOrderStatus;
import nl.tulipsolutions.api.priv.LimitOrderStatus;
import nl.tulipsolutions.api.priv.PrivateActiveOrdersServiceGrpc;
import nl.tulipsolutions.api.priv.PrivateActiveOrdersServiceGrpc.PrivateActiveOrdersServiceStub;
import nl.tulipsolutions.api.priv.StreamActiveOrdersRequest;


public class PrivateActiveOrdersServiceStreamActiveOrders {
    public static void run(ManagedChannel channel) {
        // CODEINCLUDE-BEGIN-MARKER: ref-code-example-request
        PrivateActiveOrdersServiceStub stub = PrivateActiveOrdersServiceGrpc.newStub(channel);

        // Create a request for streaming all your active orders
        // no fields are set as it does not have any
        StreamActiveOrdersRequest request = StreamActiveOrdersRequest.getDefaultInstance();

        // Make the request asynchronously
        stub.streamActiveOrders(request, new StreamObserver<ActiveOrderStatus>() {
            public void onNext(ActiveOrderStatus response) {
                System.out.println(response);
                // CODEINCLUDE-END-MARKER: ref-code-example-request
                parseAndPrint(response);
                // CODEINCLUDE-BEGIN-MARKER: ref-code-example-request
            }

            public void onError(Throwable t) {
                System.err.println("PrivateActiveOrdersService.StreamActiveOrders error: " + t.getMessage());
            }

            public void onCompleted() {
                System.out.println("PrivateActiveOrdersService.StreamActiveOrders completed");
            }
        });
        // CODEINCLUDE-END-MARKER: ref-code-example-request
    }

    public static void parseAndPrint(ActiveOrderStatus response) {
        // CODEINCLUDE-BEGIN-MARKER: ref-code-example-response
        String formattedOrderType, formattedDeadline;
        if (response.getOrderCase() == ActiveOrderStatus.OrderCase.LIMIT_ORDER) {
            LimitOrderStatus limitOrder = response.getLimitOrder();
            formattedOrderType =
                String.format(
                    "%s %f@%f remaining %f",
                    limitOrder.getSide(),
                    limitOrder.getBaseAmount(),
                    limitOrder.getPrice(),
                    limitOrder.getBaseRemaining()
                );
        } else {
            // Note that market orders do not show in active orderbook.
            formattedOrderType = "removed from orderbook";
        }
        if (response.getDeadlineNs() != 0) {
            formattedDeadline = String.format("deadline @ %d", response.getDeadlineNs());
        } else {
            formattedDeadline = "(no deadline)";
        }
        System.out.printf(
            "%s: %d for market %s %s %s\n",
            response.getClass().getSimpleName(),
            response.getOrderId(),
            response.getMarket(),
            formattedOrderType,
            formattedDeadline
        );
        // CODEINCLUDE-END-MARKER: ref-code-example-response
    }
}
